custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/restful_component.py

```diff
@@ -0,0 +1,351 @@
+"""
+The RESTfulComponent class is used to create RESTful model serving components.
+It is an abstract class that is supposed to be extended and the sub-class is
+supposed to implement its abstract methods.
+
+The class provide the core functionality of the model restful serving, which
+includes setup and preparation of the environment, configurations as well as
+executing of required services (.e.g uwsig, nginx)
+"""
+
+import abc
+import os
+import socket
+import tempfile
+import time
+import logging
+import subprocess
+import sys
+
+from mlpiper.pipeline import java_mapping
+from mlpiper.model.model_env import ModelEnv
+from mlpiper.ml_engine.rest_model_serving_engine import RestModelServingEngine
+from mlpiper.common import constants
+from mlpiper.components import parameter
+from mlpiper.common.mlpiper_exception import MLPiperException
+from mlpiper.components.connectable_component import ConnectableComponent
+from mlpiper.components.restful.constants import (
+    SharedConstants,
+    ComponentConstants,
+    UwsgiConstants,
+    NginxConstants,
+)
+from mlpiper.components.restful.constants import RestfulConstants
+from mlpiper.components.restful.nginx_broker import NginxBroker
+from mlpiper.components.restful.flask_route import FlaskRoute
+from mlpiper.components.restful.metric import Metric, MetricType, MetricRelation
+
+from datarobot_drum.resource.components.Python.uwsgi_component.uwsgi_broker import UwsgiBroker
+
+
+class RESTfulComponent(ConnectableComponent):
+    _uuid_engine = None
+    _stats_path_filename = None
+    _stats_count = 0
+
+    def __init__(self, engine):
+        # When this component is created from uwsgi entry point, engine param is None.
+        # So check for uwsgi installation only once from pipeline/component creation,
+        # when engine is not None.
+        if engine is not None:
+            if subprocess.run(
+                [sys.executable, "-m", "pip", "show", "uwsgi"],
+                stdout=subprocess.DEVNULL,
+                stderr=subprocess.DEVNULL
+            ).returncode != 0:
+                err_msg = "'uwsgi' package is required. " \
+                          "If you are running on a Windows system, " \
+                          "you may not be able to use RESTful component. " \
+                          "Check how to install uwsgi on Windows. " \
+                          "To install uwsgi on a Mac/Linux system, call " \
+                          "`pip install uwsgi` or `pip install mlpiper[uwsgi]`"
+                raise MLPiperException(err_msg)
+
+        super(RESTfulComponent, self).__init__(
+            engine if engine else RestModelServingEngine("uwsgi-context")
+        )
+        self._dry_run = False
+        self._wsgi_broker = None
+        self._nginx_broker = None
+        self._wid = None
```

Comment:
what is `wid` ? and `wuuid` ?

Reply:
wid is uwsgi.worker_id() 
wuuid  is str(uuid.uuid4()) just uniq id